### PR TITLE
Fix sub dependencies

### PR DESCRIPTION
**Pull request by dcbaker**
As @mensinda pointed out.

### CODE REVIEW

**Comment by dcbaker** [comment]
@jpakkane, @nirbheek, either of you want to review this?

**Comment by dcbaker** [comment]
Code duplication. There are at least three places where we have to initialize ThreadDependency internally (and I missed one!) that need this code.

**Comment by nirbheek** [comment]
Hmm, I'm not sure I understand. I can only see three places where `ThreadDependency` is *instantiated*. Wouldn't this code also work?

```diff
 class ThreadDependency(ExternalDependency):
-    def __init__(self, environment, kwargs):
+    def __init__(self, environment: 'Environment', kwargs: typing.Any):
+        kwargs['method'] = 'auto'
         super().__init__('threads', environment, None, kwargs)
```

**Comment by dcbaker** [comment]
I guess we could put it in the default constructor, I felt like leaving that alone was probably nicer, but I'll remove the init_internal method.

**Comment by dcbaker** [comment]
The more I think about doing it this way the more uncomfortable it makes me. doing this in the default constructor of `ThreadDependency` works because it accepts only auto. But imagine if there was some dependency that needed to initialize another dependency internally that did take multiple methods, porting your change to that code (say boost) would effectively remove the 'method' keyword argument from that dependency. I can drop the specialized constructor if you'd like, but I really, really don't want to set `kwargs['method']` inside the `ThreadDependency` constructor.

File: mesonbuild/dependencies/misc.py [hunk 1]
```diff
@@ -384,13 +388,24 @@ def __init__(self, environment, kwargs):
 
 
 class ThreadDependency(ExternalDependency):
-    def __init__(self, environment, kwargs):
+    def __init__(self, environment: 'Environment', kwargs: typing.Any):
```
Comment (mensinda) [inline mesonbuild/dependencies/misc.py]:
```suggestion
    def __init__(self, environment: 'Environment', kwargs: typing.Any):
        assert(kwargs['__ThreadDependency_custom_init__'] == True)
```

File: mesonbuild/dependencies/misc.py [hunk 2]
```diff
@@ -384,13 +388,24 @@ def __init__(self, environment, kwargs):
...
 
+    @classmethod
+    def internal_init(cls, environment: 'Environment', kwargs: typing.Any) -> 'ThreadDependency':
+        """A constructor for dependencies to create a dependency on Threads.
+
+        This differs from the main constructor only in that it strips
+        the 'method' kwarg
+        """
+        kwargs = kwargs.copy()
+        kwargs['method'] = 'auto'
```
Comment (mensinda) [inline mesonbuild/dependencies/misc.py]:
```suggestion
        kwargs['method'] = 'auto'
        kwargs['__ThreadDependency_custom_init__'] = True
```

**Comment by dcbaker** [comment]
Maybe in that case it would be better to have a special function that we use internally, something like:

```python
class Dependency:
    ...
    def add_new_dependency(self, dep: typing.Type[Dependency], args: List[Any], kwargs: Dict[str, Any]) -> None:
        kwargs = kwargs.copy()
        delete kwargs['method']
        self.deps.append(dep(*args, **kwargs))
```
which would be a more general solution.

**Comment by mensinda** [comment]
Sure, a central function would be an even better solution. However, this function should also be aware of cases where deleting the `method` is **not** desirable.

**Comment by jpakkane** [comment]
I have no idea what should be happening here but feel free to merge once everything is ok.

File: mesonbuild/dependencies/base.py [hunk 3]
```diff
@@ -177,6 +180,20 @@ def get_partial_dependency(self, *, compile_args: bool = False,
         """
         raise RuntimeError('Unreachable code in partial_dependency called')
 
+    def _add_sub_dependency(self, dep_type: typing.Type[Dependency], env: Environment,
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ [F821] undefined name 'Dependency' ([link](https://sider.review/gh/repos/19784232/pulls/5254#issue-4011778))

<sub>posted by Sider</sub>

File: mesonbuild/dependencies/base.py [hunk 4]
```diff
@@ -177,6 +180,20 @@ def get_partial_dependency(self, *, compile_args: bool = False,
...
+        """Add an internal dependency of of the given type.
+
+        This method is intended to simplify cases of adding a dependency on
+        another dependency type (such as threads). This will by default set
+        the method back to auto, but the 'method' keyword argument can be
+        used to overwrite this behavior.
+        """
+        kwargs = kwargs.copy()
+        kwargs['method'] = method
+        self.deps.append(dep_type(*args, **kwargs))
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ [F821] undefined name 'args' ([link](https://sider.review/gh/repos/19784232/pulls/5254#issue-4011779))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
@mensinda, does this approach look good to you? I tested your cmake llvm series and that seems to work so if this is good I'll merge this and the cmake llvm

**Comment by mensinda** [comment]
Yes the current apporach looks good to me. I will rebase my CMake LLVM PR once this gets merged.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

**Comment by dcbaker** [comment]
I'm planning to merge this then as long as the CI comes back green.

**Pull request merged**